pyop2/fusion.py

```diff
@@ -0,0 +1,1264 @@
+
+    @staticmethod
+    def specialize(args, gtl_map, loop_id):
+        """Given ``args``, instances of some :class:`fusion.Arg` superclass,
+        create and return specialized :class:`fusion.Arg` objects.
+
+        :param args: either a single :class:`host.Arg` object or an iterator
+                     (accepted: list, tuple) of :class:`host.Arg` objects.
+        :gtl_map: a dict associating global maps' names to local maps' c_names.
+        :param loop_id: indicates the position of the args` loop in the loop
+                        chain
+        """
+
+        def convert(arg, gtl_map, loop_id):
+            # Retrive local maps
+            maps = as_tuple(arg.map, Map)
+            c_local_maps = [None]*len(maps)
+            for i, map in enumerate(maps):
+                c_local_maps[i] = [None]*len(map)
+                for j, m in enumerate(map):
+                    c_local_maps[i][j] = gtl_map["%s%d_%d" % (m.name, i, j)]
+            # Instantiate and initialize new, specialized Arg
+            _arg = Arg(arg.data, arg.map, arg.idx, arg.access, arg._flatten)
+            _arg._loop_position = loop_id
+            _arg._position = arg._position
+            _arg._indirect_position = arg._indirect_position
+            _arg._c_local_maps = c_local_maps
+            return _arg
+
+        if isinstance(args, (list, tuple)):
+            return [convert(arg, gtl_map, loop_id) for arg in args]
+        return convert(args, gtl_map, loop_id)
+
+    @staticmethod
+    def filter_args(loop_args):
+        """Given a sequence of tuples of ``Args``, where each tuple comes from a
+        different loop, create a sequence of ``Args`` where there are no duplicates
+        and access modes are properly set (for example, an ``Arg`` whose ``Dat``
+        appears in two different tuples with access mode ``WRITE`` and ``READ``,
+        respectively, will have access mode ``RW`` in the returned sequence of
+        ``Args``."""
+        filtered_args = OrderedDict()
+        for args in loop_args:
+            for a in args:
+                filtered_args[a.data] = filtered_args.get(a.data, a)
+                if a.access != filtered_args[a.data].access:
+                    if READ in [a.access, filtered_args[a.data].access]:
+                        # If a READ and some sort of write (MIN, MAX, RW, WRITE,
+                        # INC), then the access mode becomes RW
+                        filtered_args[a.data]._access = RW
+                    elif WRITE in [a.access, filtered_args[a.data].access]:
+                        # Can't be a READ, so just stick to WRITE regardless of what
+                        # the other access mode is
+                        filtered_args[a.data]._access = WRITE
+                    else:
+                        # Neither READ nor WRITE, so access modes are some
+                        # combinations of RW, INC, MIN, MAX. For simplicity,
+                        # just make it RW.
```

Comment:
Yes, but look at this line 107; a new access mode is set *iff* the access modes of two arguments with same payload are different. So, if they are both ``INC``, the access mode actually remains ``INC``.  

The comment "For simplicity..." is just to say "instead of having the product of all of the possible access modes (RW_INC, RW_MIN, etc.), which would probably be useless, let's just stick to the general one, RW"